bigtable/google/cloud/bigtable/row_data.py

```diff
@@ -284,7 +309,36 @@ def state(self):
 
     def cancel(self):
         """Cancels the iterator, closing the stream."""
-        self._response_iterator.cancel()
+        self.response_iterator.cancel()
+
+    def _create_retry_request(self):
+        """Helper for :meth:`read_rows`."""
+        row_range = self.request.rows.row_ranges.pop()
+        range_kwargs = {}
+        # start AFTER the row_key of the last successfully read row
+        range_kwargs['start_key_open'] = self.last_scanned_row_key
+        range_kwargs['end_key_open'] = row_range.end_key_open
+        self.request.rows.row_ranges.add(**range_kwargs)
+
+    def _on_error(self, exc):
+        """Helper for :meth:`read_rows`."""
+        # restart the read scan from AFTER the last successfully read row
+        if self.last_scanned_row_key:
+            self._create_retry_request()
+
+        self.response_iterator = self.read_method(self.request)
+
+    def _read_next(self):
+        """Helper for :meth:`read_rows`."""
+        return six.next(self.response_iterator)
+
+    def _read_next_response(self):
+        """Helper for :meth:`read_rows`."""
+        next_response = functools.partial(self._read_next)
```

Comment:
I don't think you actually need to use `functools.partial` here anymore.

Reply:
Thanks! Your interface recommendations made a big improvement to the design.  Also, I think we finally understand the full range of features in Retry. For this implementation we needed ```on_error```.

We will take a look at the last review comment about partial and see if this can come out.